glean-core/src/lib.rs

```diff
@@ -500,6 +508,13 @@ pub fn shutdown() {
         log::error!("Can't shutdown dispatcher thread: {:?}", e);
     }
 
+    // Call on_shutdown.
+    // Need to be done without holding a lock on the global Glean object.
+    {
+        let state = global_state().lock().unwrap();
+        state.callbacks.on_shutdown();
```

Comment:
If we're worried about timeouts, we could spawn a thread for `on_shutdown` and give it one half of a Condvar that this thread'll be `wait_timeout()`-ing on?

Reply:
That'd solve the arbitrarily long part, and if we named the thread (or used some other thread-local storage) we could have the dispatcher forbid adding of tasks from that thread, which'd solve the "do not add tasks" part.

Reply:
Both implemented now.